### PR TITLE
fix(TDI-42653): reduce REST API calls

### PR DESCRIPTION
**Pull request by undx**
**What is the current behavior?** (You can also link to an open issue here)

https://jira.talendforge.org/browse/TDI-42653

**What is the new behavior?**
Assign service client when connection is validated to prevent another API call to get an already acquired access token.
Provide a class to cache and share acquired access tokens.

**Please check if the PR fulfills these requirements**

- [X] The commit message follows Talend standard
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?
- [X] The code coverage on new code >75%
- [X] The new code does not introduce new technical issues (sonar / eslint)

**What kind of change does this PR introduce?**

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [X] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/23/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/23/) 
[Sonar results](https://sonar.datapwn.com/overview?id=org.talend.components%3Acomponents%3Aundx%2FTDI-42653_MarketoUniqRestCall)

**Comment by build-talend-common** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/24/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/24/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/25/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/25/) 
[Sonar results](https://sonar.datapwn.com/overview?id=org.talend.components%3Acomponents%3Aundx%2FTDI-42653_MarketoUniqRestCall)

File: components/components-marketo/components-marketo-runtime/src/main/java/org/talend/components/marketo/runtime/MarketoAccessTokenPool.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    public static MarketoAccessTokenPool getInstance() {
+        return ourInstance;
+    }
+
+    private MarketoAccessTokenPool() {
+        tokens = new HashMap<>();
+    }
+
+    public void invalidateToken(Integer connectionHash){
+        tokens.put(connectionHash, null);
```
Comment (chmyga) [inline components/components-marketo/components-marketo-runtime/src/main/java/org/talend/components/marketo/runtime/MarketoAccessTokenPool.java]: better to remove the token from map than put null

Comment (undx) [reply components/components-marketo/components-marketo-runtime/src/main/java/org/talend/components/marketo/runtime/MarketoAccessTokenPool.java]: right. done

**Comment by undx** [comment]
> instance could be initialized lazily, but that is not critical. It looks ok if MarketoAccessTokenPool instance is not going to be used in multithreaded environment. Otherwise, the code should be rewritten.

That's not the case so should be ok.

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/30/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/30/) 
[Sonar results](https://sonar.datapwn.com/overview?id=org.talend.components%3Acomponents%3Aundx%2FTDI-42653_MarketoUniqRestCall)

File: components/components-marketo/components-marketo-runtime/src/main/java/org/talend/components/marketo/runtime/client/MarketoBaseRESTClient.java [hunk 2]
```diff
@@ -153,6 +156,7 @@ public MarketoBaseRESTClient(TMarketoConnectionProperties connection) throws Mar
         endpoint = connection.endpoint.getValue();
         userId = connection.clientAccessId.getValue();
         secretKey = connection.secretKey.getValue();
+        connectionHash = (endpoint + userId + secretKey).hashCode();
```
Comment (chmyga) [inline components/components-marketo/components-marketo-runtime/src/main/java/org/talend/components/marketo/runtime/client/MarketoBaseRESTClient.java]: in some cases we might have collisions in hash codes. Why don't we use some kind of randomly generated UID for this case? or, maybe, at least endpoint + hash as a token identifier?

Comment (undx) [reply components/components-marketo/components-marketo-runtime/src/main/java/org/talend/components/marketo/runtime/client/MarketoBaseRESTClient.java]: The tokens should be accessed by an id that is based on connection infos, using a random uid cannot allow the get it back from unless stored. Made some tests using hashcode and it should be ok. I won't expect collisions.

**Pull request merged**